src/lxc/lxccontainer.c

```diff
@@ -4691,6 +4712,119 @@ static bool do_lxcapi_restore(struct lxc_container *c, char *directory, bool ver
 
 WRAP_API_2(bool, lxcapi_restore, char *, bool)
 
+static int do_lxcapi_mount(struct lxc_container *c,
+			const char *source, const char *target,
+			const char *filesystemtype, unsigned long mountflags,
+			const void *data, struct lxc_mount *mnt) {
+	char *suff, *sret;
+	char template[MAXPATHLEN], path[MAXPATHLEN];
+	pid_t pid, init_pid;
+	size_t len;
+	int ret = -1, fd = -EBADF;
+
+	if (!c || !c->lxc_conf) {
+		ERROR("Container or configuration is NULL");
+		return -EINVAL;
+	}
+
+	if (!c->lxc_conf->lxc_shmount.path_host) {
+		ERROR("Host path to shared mountpoint must be specified in the config\n");
+		return -EINVAL;
+	}
+	len = strlen(c->lxc_conf->lxc_shmount.path_host) + sizeof("/.lxcmount_XXXXXX") - 1;
+
+	ret = snprintf(template, len + 1, "%s/.lxcmount_XXXXXX", c->lxc_conf->lxc_shmount.path_host);
+	if (ret < 0 || (size_t)ret >= len + 1) {
+		SYSERROR("Error writing shmounts tempdir name");
+		goto out;
+	}
+
+	/* Create a temporary dir under the shared mountpoint */
+	sret = mkdtemp(template);
```

Comment:
I just realized that your API addition will currently only work for directory-based mounts. Here you need to create either a temporary directory or a temporary file depending on whether the source of the mount is a file or a directory.
Your tests don't check for bind-mounting a file into a container. The test that looks like it does it is:
```
ret = c->mount(c, "/dev/urandom", "/mnt/mount_injection_test_urandom", "devtmpfs", 0, NULL, &mnt);
```
But it's not mounting a device actually since your telling the kernel to mount `devtmpfs` which is a directory-based mount. That means the kernel will also ignore the `/dev/urandom` source path. It will be listed as the source but it actually isn't. :)

The test that you likely want in addition to the `devtmpfs` one after changing the new API function is:
```
ret = c->mount(c, "/dev/urandom", "/mnt/mount_injection_test_urandom", NULL, MS_BIND, NULL, &mnt);
```

Reply:
You can use 

https://github.com/lxc/lxc/blob/9ab207cad74f77ce643d98334107949ade97463b/src/lxc/utils.c#L2388
to create the temporary file.

Reply:
Also, I would prefer it if the temporary directory or temporary file you're creating as the mount target would have permission `0000` instead of `0700` and `0600` since we don't need any perms at all. :) Unless this doesn't work.

Reply:
Somehow I didn't find any analogues of `mkdir_p()` for files so I adopted some nice utils from LXD code [here](https://github.com/lxc/lxd/blob/master/lxd/main_forkmount.go#L51) (and also their logic regarding permissions of the files being created inside the container with 0755).